File: apps/web/lib/twelveLifeStages.ts
/**
 * Twelve Life Stages (长生十二宫 / Chang Sheng 12 Stages)
 *
 * The lunar-javascript library computes the life stage for each pillar via
 * bazi.getYearDiShi(), getMonthDiShi(), getDayDiShi(), getTimeDiShi().
 * This module enriches those raw Chinese strings with English translations,
 * and provides a manual 自坐 calculation using each pillar's own stem.
 */

export interface LifeStageInfo {
  chinese: string;
  english: string;
}

const LIFE_STAGE_INFO: Record<string, string> = {
  '长生': 'Birth',
  '沐浴': 'Bathing',
  '冠带': 'Crowning',
  '临官': 'Official',
  '帝旺': 'Peak',
  '衰':   'Decline',
  '病':   'Illness',
  '死':   'Death',
  '墓':   'Tomb',
  '绝':   'Void',
  '胎':   'Embryo',
  '养':   'Nourishment',
};

/**
 * Enrich a raw Chinese life stage string (from lunar-javascript) with its
 * English translation.
 *
 * @param chinese  The Chinese stage name returned by bazi.getYearDiShi() etc.
 * @returns LifeStageInfo or null if the value is unrecognised.
 */
export function getLifeStageInfo(chinese: string): LifeStageInfo | null {
  const english = LIFE_STAGE_INFO[chinese];
  if (!english) return null;
  return { chinese, english };
}

// ─── 自坐 (Self-Seated) Life Stage Calculation ───────────────────────────────
//
// The library's _getDiShi() is hardcoded to the Day Master stem. For 自坐 we
// replicate the same algorithm using each pillar's own Heavenly Stem instead.
// LunarUtil is exported but its tables use i18n placeholder keys ({tg.jia} etc.)
// rather than Chinese characters, so we encode the offsets directly here.

// The 12 stages in index order (mirrors LunarUtil.CHANG_SHENG)
const CHANG_SHENG_ORDER = [
  '长生','沐浴','冠带','临官','帝旺','衰','病','死','墓','绝','胎','养',
] as const;

// Starting offset per Heavenly Stem (mirrors LunarUtil.CHANG_SHENG_OFFSET)
// Yang stems progress clockwise (+zhiIndex), Yin stems counter-clockwise (-zhiIndex)
// Earth stems (戊/己) share the same offset as Fire stems (丙/丁) — standard BaZi rule
const CHANG_SHENG_OFFSET: Record<string, number> = {
  '甲': 1, '丙': 10, '戊': 10, '庚': 7, '壬': 4,
  '乙': 6, '丁':  9, '己':  9, '辛': 0, '癸': 3,
};

// Earthly Branch to 0-based index (子=0 … 亥=11)
const ZHI_INDEX: Record<string, number> = {
  '子': 0, '丑': 1, '寅': 2, '卯': 3, '辰': 4, '巳': 5,
  '午': 6, '未': 7, '申': 8, '酉': 9, '戌': 10, '亥': 11,
};

const YANG_STEMS = new Set(['甲', '丙', '戊', '庚', '壬']);

/**
 * Compute the 自坐 (self-seated) life stage for a pillar.
 * Uses the pillar's own Heavenly Stem as the reference against its own Earthly Branch,
 * replicating the library's _getDiShi() algorithm with a custom stem.
 *
 * @param stem    Heavenly Stem of the pillar (e.g. "甲")
 * @param branch  Earthly Branch of the pillar (e.g. "子")
 * @returns LifeStageInfo or null if stem/branch are unrecognised.
 */
export function getPillarLifeStageInfo(stem: string, branch: string): LifeStageInfo | null {
  const offset = CHANG_SHENG_OFFSET[stem];
  const zhiIndex = ZHI_INDEX[branch];
  if (offset === undefined || zhiIndex === undefined) return null;
  const isYang = YANG_STEMS.has(stem);
  const rawIndex = offset + (isYang ? zhiIndex : -zhiIndex);
  const index = ((rawIndex % 12) + 12) % 12;
  return getLifeStageInfo(CHANG_SHENG_ORDER[index]);
}

File: apps/backend/astronomer_logic/twelve_life_stages.py
"""
十二长生 Twelve Life Stages (Di Shi)

Two values per pillar (年柱, 月柱, 日柱, 时柱):
  - library_stage:     raw Chinese string from the lunar-python library
                       (bazi.getYearDiShi() etc.) — referenced against the Day Master.
  - self_seated_stage: 自坐 — computed from the pillar's OWN 天干 against its own 地支,
                       using the same Chang Sheng offset algorithm as twelveLifeStages.ts.

Self-seated algorithm:
  offset = CHANG_SHENG_OFFSET[天干]
  index  = (offset + zhi_index)  for yang stems (甲丙戊庚壬)
           (offset - zhi_index)  for yin stems  (乙丁己辛癸)
  stage  = CHANG_SHENG_ORDER[index % 12]
"""

# 12 stages in index order (mirrors LunarUtil.CHANG_SHENG)
_CHANG_SHENG_ORDER = [
    '长生', '沐浴', '冠带', '临官', '帝旺',
    '衰',   '病',   '死',   '墓',   '绝',   '胎', '养',
]

# Starting offset per Heavenly Stem
# Earth stems (戊/己) share Fire stems' offsets — standard BaZi rule
_CHANG_SHENG_OFFSET = {
    '甲': 1,  '丙': 10, '戊': 10, '庚': 7, '壬': 4,
    '乙': 6,  '丁':  9, '己':  9, '辛': 0, '癸': 3,
}

# Earthly Branch → 0-based index (子=0 … 亥=11)
_ZHI_INDEX = {
    '子': 0, '丑': 1, '寅': 2, '卯': 3, '辰': 4,  '巳': 5,
    '午': 6, '未': 7, '申': 8, '酉': 9, '戌': 10, '亥': 11,
}

_YANG_STEMS = {'甲', '丙', '戊', '庚', '壬'}


def _self_seated_stage(stem: str, branch: str) -> str | None:
    """Compute the 自坐 (self-seated) life stage using the pillar's own 天干 and 地支."""
    offset = _CHANG_SHENG_OFFSET.get(stem)
    zhi_index = _ZHI_INDEX.get(branch)
    if offset is None or zhi_index is None:
        return None
    raw = offset + (zhi_index if stem in _YANG_STEMS else -zhi_index)
    return _CHANG_SHENG_ORDER[raw % 12]


def get_twelve_life_stages(bazi, pillars: dict) -> dict:
    """
    Return the two Di Shi life-stage values for each of the Four Pillars.

    Args:
        bazi:    EightChar object from lunar_birthday.getEightChar()
        pillars: Output of get_bazi_pillars() — provides 天干 and 地支 for self-seated calculation.

    Returns:
        dict keyed by 年柱, 月柱, 日柱, 时柱.
        Each contains: library_stage (str), self_seated_stage (str | None).
    """
    def make_stage(library_fn, pillar_key: str) -> dict:
        stem   = pillars[pillar_key]["天干"]
        branch = pillars[pillar_key]["地支"]
        return {
            "星运":     library_fn(),
            "自坐": _self_seated_stage(stem, branch),
        }

    return {
        "年柱": make_stage(bazi.getYearDiShi,  "年柱"),
        "月柱": make_stage(bazi.getMonthDiShi, "月柱"),
        "日柱": make_stage(bazi.getDayDiShi,   "日柱"),
        "时柱": make_stage(bazi.getTimeDiShi,  "时柱"),
    }
